scripts/gas-refund-program/transactions-validation/validateTransactions.ts

```diff
@@ -62,55 +63,89 @@ export async function validateTransactions() {
         'status',
         'refundedAmountPSP',
         'refundedAmountUSD',
+        'pspUsd',
       ],
     });
 
     if (!transactionsSlice.length) break;
 
     offset += pageSize;
 
-    const transactionsWithUpdatedStatus = [];
+    const updatedTransactions = [];
 
     for (const tx of transactionsSlice) {
-      let newStatus;
-
-      const isGlobalLimitReached =
-        guardian.state.totalPSPRefunded
-          .plus(tx.refundedAmountPSP)
-          .isGreaterThan(MAX_PSP_GLOBAL_BUDGET) ||
-        guardian.isMaxPSPGlobalBudgetSpent();
-
-      const isLocalLimitReached =
-        guardian
-          .totalRefundedAmountUSD(tx.address)
-          .plus(tx.refundedAmountUSD)
-          .isGreaterThan(MAX_USD_ADDRESS_BUDGET) ||
-        guardian.isAccountUSDBudgetSpent(tx.address);
-
-      // once one limit is reached, reject tx. Note: this would let refund future transactions that are still within the limit
-      if (isLocalLimitReached || isGlobalLimitReached) {
-        newStatus = TransactionStatus.REJECTED;
+      let { status } = tx;
+      let refundedAmountPSP = new BigNumber(tx.refundedAmountPSP);
+      let refundedAmountUSD = new BigNumber(tx.refundedAmountUSD);
+
+      const isGlobalLimitReached = guardian.isMaxPSPGlobalBudgetSpent();
+      const isPerAccountLimitReached = guardian.isAccountUSDBudgetSpent(
+        tx.address,
+      );
+
+      if (isGlobalLimitReached || isPerAccountLimitReached) {
+        status = TransactionStatus.REJECTED;
       } else {
-        newStatus = TransactionStatus.VALIDATED;
+        status = TransactionStatus.VALIDATED;
+
+        if (
+          guardian.state.totalPSPRefunded
+            .plus(tx.refundedAmountPSP)
+            .isGreaterThan(MAX_PSP_GLOBAL_BUDGET)
+        ) {
+          // Note: updating refundedAmountUSD does not matter if global budget limit is reached
```

Comment:
edge case: tx surpasses both program budget and address budget. In that case we would limit the psp amount relative to the global budget, but that amount could still put the user over 30k (eg before tx we were $20 away from program budget and 10$ away from address budget. the tx qualifies for $50 gas refund - at present the user would get $20 taking them to $30,010).
We could address this by defining three vars:
```
isOverProgramBudget = ...
isOverAddressBudget = ...
isOverBothProgramAndAddressBudget = isOverProgramBudget && isOverAddressBudget
```
and then handle them in a `switch` (or just three `if`s). For `isOverProgramAndUserBudget` we could calculate the  `refundedAmountPSP` for both cases (surpassing the program budget, and the address budget) then use the minimum (while also setting the usd amount based of that).

Reply:
This has been mitigated but can be addressed we're closer to the limit